<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>asymmetree.seqevolve.IndelModel API documentation</title>
<meta name="description" content="Indel models for sequence evolution â€¦" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>asymmetree.seqevolve.IndelModel</code></h1>
</header>
<section id="section-intro">
<p>Indel models for sequence evolution.</p>
<h2 id="references">References</h2>
<p>.. [1] Z. Yang.
Computational molecular evolution.
Oxford series in ecology and evolution. Oxford University Press, 2006.
ISBN 978-0-19-856699-1 978-0-19-856702-8.
.. [2] R. A. Cartwright.
DNA assembly with gaps (Dawg): Simulating sequence evolution.
In: Bioinformatics, 21(Suppl 3):iii31-iii38, November 2005.
doi:10.1093/bioinformatics/bti1200.</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># -*- coding: utf-8 -*-

&#34;&#34;&#34;Indel models for sequence evolution.

References
----------
.. [1] Z. Yang.
   Computational molecular evolution.
   Oxford series in ecology and evolution. Oxford University Press, 2006.
   ISBN 978-0-19-856699-1 978-0-19-856702-8.
.. [2] R. A. Cartwright.
   DNA assembly with gaps (Dawg): Simulating sequence evolution.
   In: Bioinformatics, 21(Suppl 3):iii31-iii38, November 2005.
   doi:10.1093/bioinformatics/bti1200.
&#34;&#34;&#34;

from asymmetree.tools.Sampling import Sampler


__author__ = &#39;David Schaller&#39;


class IndelModel:
    &#34;&#34;&#34;Indel model&#34;&#34;&#34;
    
    def __init__(self, insertion_rate, deletion_rate,
                 length_distr=(&#39;zipf&#39;, 1.821),        # (Chang and Benner 2004)
                 min_length=1,
                 max_length=None,
                 **kwargs):
        &#34;&#34;&#34;
        Parameters
        ----------
        insertition_rate : float
            Insertion rate w.r.t. units of evolutionary distance.
        deletion_rate : float
            Deletion rate w.r.t. units of evolutionary distance.
        length_distr : tuple, optional
            Distribution of the length of the indels, see documentation for
            available options. The default is a Zipf distribution as observed
            for empirical data in [1].
        min_length : int, optional
            The minimal length of an indel, default is 1.
        max_length : int, optional
            The maximal length of an indel. The default is None in which case
            the distribution is not truncated (on the upper side).
        
        References
        ----------
        .. [1] M. S. Chang and S. A. Benner.
           Empirical Analysis of Protein Insertions and Deletions Determining
           Parameters for the Correct Placement of Gaps in Protein Sequence
           Alignments. 
           In: Journal of Molecular Biology, 341(2):617-631, August 2004.
           doi:10.1016/j.jmb.2004.05.045.
        &#34;&#34;&#34;
        
        if insertion_rate &lt; 0.0 or deletion_rate &lt; 0.0:
            raise ValueError(&#34;insertion and deletion rates must be non-negative&#34;)
        else:
            self._ins_rate = insertion_rate     # insertion rate per site
            self._del_rate = deletion_rate      # deletion rate per site
        
        if isinstance(length_distr, Sampler):
            self.sampler = length_distr
        else:
            self.sampler = Sampler(length_distr,
                                   maximum=max_length,
                                   discrete=True)
        
    
    def get_rates(self, seq_length):
        &#34;&#34;&#34;Return the current insertion and deletion rate.
        
        Computes the current insertion and deletion rate according to the model
        and the length of the sequence.
        
        Parameters
        ----------
        seq_length : int
            The length of the sequence.
        
        Returns
        -------
        tuple of two floats
            The total rate for an insertation and deletion event, see [1].
            
        References
        ----------
        .. [1] R. A. Cartwright.
           DNA assembly with gaps (Dawg): Simulating sequence evolution.
           In: Bioinformatics, 21(Suppl 3):iii31-iii38, November 2005.
           doi:10.1093/bioinformatics/bti1200.
        &#34;&#34;&#34;
        
        # expected value may be infinite for zipf distribution, in which case
        # the minimal length is used, see &#39;Sampler&#39;
        return ( (seq_length + 1) * self._ins_rate,
                 (seq_length + self.sampler._exp_val - 1) * self._del_rate )
        
    
    def draw_length(self):
        &#34;&#34;&#34;Draw the length for an indel.
        
        Returns
        -------
        int
            The drawn length for the indel.
        &#34;&#34;&#34;
        
        return self.sampler()
    </code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="asymmetree.seqevolve.IndelModel.IndelModel"><code class="flex name class">
<span>class <span class="ident">IndelModel</span></span>
<span>(</span><span>insertion_rate, deletion_rate, length_distr=('zipf', 1.821), min_length=1, max_length=None, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Indel model</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>insertition_rate</code></strong> :&ensp;<code>float</code></dt>
<dd>Insertion rate w.r.t. units of evolutionary distance.</dd>
<dt><strong><code>deletion_rate</code></strong> :&ensp;<code>float</code></dt>
<dd>Deletion rate w.r.t. units of evolutionary distance.</dd>
<dt><strong><code>length_distr</code></strong> :&ensp;<code>tuple</code>, optional</dt>
<dd>Distribution of the length of the indels, see documentation for
available options. The default is a Zipf distribution as observed
for empirical data in [1].</dd>
<dt><strong><code>min_length</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>The minimal length of an indel, default is 1.</dd>
<dt><strong><code>max_length</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>The maximal length of an indel. The default is None in which case
the distribution is not truncated (on the upper side).</dd>
</dl>
<h2 id="references">References</h2>
<p>.. [1] M. S. Chang and S. A. Benner.
Empirical Analysis of Protein Insertions and Deletions Determining
Parameters for the Correct Placement of Gaps in Protein Sequence
Alignments.
In: Journal of Molecular Biology, 341(2):617-631, August 2004.
doi:10.1016/j.jmb.2004.05.045.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class IndelModel:
    &#34;&#34;&#34;Indel model&#34;&#34;&#34;
    
    def __init__(self, insertion_rate, deletion_rate,
                 length_distr=(&#39;zipf&#39;, 1.821),        # (Chang and Benner 2004)
                 min_length=1,
                 max_length=None,
                 **kwargs):
        &#34;&#34;&#34;
        Parameters
        ----------
        insertition_rate : float
            Insertion rate w.r.t. units of evolutionary distance.
        deletion_rate : float
            Deletion rate w.r.t. units of evolutionary distance.
        length_distr : tuple, optional
            Distribution of the length of the indels, see documentation for
            available options. The default is a Zipf distribution as observed
            for empirical data in [1].
        min_length : int, optional
            The minimal length of an indel, default is 1.
        max_length : int, optional
            The maximal length of an indel. The default is None in which case
            the distribution is not truncated (on the upper side).
        
        References
        ----------
        .. [1] M. S. Chang and S. A. Benner.
           Empirical Analysis of Protein Insertions and Deletions Determining
           Parameters for the Correct Placement of Gaps in Protein Sequence
           Alignments. 
           In: Journal of Molecular Biology, 341(2):617-631, August 2004.
           doi:10.1016/j.jmb.2004.05.045.
        &#34;&#34;&#34;
        
        if insertion_rate &lt; 0.0 or deletion_rate &lt; 0.0:
            raise ValueError(&#34;insertion and deletion rates must be non-negative&#34;)
        else:
            self._ins_rate = insertion_rate     # insertion rate per site
            self._del_rate = deletion_rate      # deletion rate per site
        
        if isinstance(length_distr, Sampler):
            self.sampler = length_distr
        else:
            self.sampler = Sampler(length_distr,
                                   maximum=max_length,
                                   discrete=True)
        
    
    def get_rates(self, seq_length):
        &#34;&#34;&#34;Return the current insertion and deletion rate.
        
        Computes the current insertion and deletion rate according to the model
        and the length of the sequence.
        
        Parameters
        ----------
        seq_length : int
            The length of the sequence.
        
        Returns
        -------
        tuple of two floats
            The total rate for an insertation and deletion event, see [1].
            
        References
        ----------
        .. [1] R. A. Cartwright.
           DNA assembly with gaps (Dawg): Simulating sequence evolution.
           In: Bioinformatics, 21(Suppl 3):iii31-iii38, November 2005.
           doi:10.1093/bioinformatics/bti1200.
        &#34;&#34;&#34;
        
        # expected value may be infinite for zipf distribution, in which case
        # the minimal length is used, see &#39;Sampler&#39;
        return ( (seq_length + 1) * self._ins_rate,
                 (seq_length + self.sampler._exp_val - 1) * self._del_rate )
        
    
    def draw_length(self):
        &#34;&#34;&#34;Draw the length for an indel.
        
        Returns
        -------
        int
            The drawn length for the indel.
        &#34;&#34;&#34;
        
        return self.sampler()</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="asymmetree.seqevolve.IndelModel.IndelModel.draw_length"><code class="name flex">
<span>def <span class="ident">draw_length</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>Draw the length for an indel.</p>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>int</code></dt>
<dd>The drawn length for the indel.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def draw_length(self):
    &#34;&#34;&#34;Draw the length for an indel.
    
    Returns
    -------
    int
        The drawn length for the indel.
    &#34;&#34;&#34;
    
    return self.sampler()</code></pre>
</details>
</dd>
<dt id="asymmetree.seqevolve.IndelModel.IndelModel.get_rates"><code class="name flex">
<span>def <span class="ident">get_rates</span></span>(<span>self, seq_length)</span>
</code></dt>
<dd>
<div class="desc"><p>Return the current insertion and deletion rate.</p>
<p>Computes the current insertion and deletion rate according to the model
and the length of the sequence.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>seq_length</code></strong> :&ensp;<code>int</code></dt>
<dd>The length of the sequence.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>tuple</code> of <code>two floats</code></dt>
<dd>The total rate for an insertation and deletion event, see [1].</dd>
</dl>
<h2 id="references">References</h2>
<p>.. [1] R. A. Cartwright.
DNA assembly with gaps (Dawg): Simulating sequence evolution.
In: Bioinformatics, 21(Suppl 3):iii31-iii38, November 2005.
doi:10.1093/bioinformatics/bti1200.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_rates(self, seq_length):
    &#34;&#34;&#34;Return the current insertion and deletion rate.
    
    Computes the current insertion and deletion rate according to the model
    and the length of the sequence.
    
    Parameters
    ----------
    seq_length : int
        The length of the sequence.
    
    Returns
    -------
    tuple of two floats
        The total rate for an insertation and deletion event, see [1].
        
    References
    ----------
    .. [1] R. A. Cartwright.
       DNA assembly with gaps (Dawg): Simulating sequence evolution.
       In: Bioinformatics, 21(Suppl 3):iii31-iii38, November 2005.
       doi:10.1093/bioinformatics/bti1200.
    &#34;&#34;&#34;
    
    # expected value may be infinite for zipf distribution, in which case
    # the minimal length is used, see &#39;Sampler&#39;
    return ( (seq_length + 1) * self._ins_rate,
             (seq_length + self.sampler._exp_val - 1) * self._del_rate )</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul>
<li><a href="#references">References</a></li>
</ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="asymmetree.seqevolve" href="index.html">asymmetree.seqevolve</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="asymmetree.seqevolve.IndelModel.IndelModel" href="#asymmetree.seqevolve.IndelModel.IndelModel">IndelModel</a></code></h4>
<ul class="">
<li><code><a title="asymmetree.seqevolve.IndelModel.IndelModel.draw_length" href="#asymmetree.seqevolve.IndelModel.IndelModel.draw_length">draw_length</a></code></li>
<li><code><a title="asymmetree.seqevolve.IndelModel.IndelModel.get_rates" href="#asymmetree.seqevolve.IndelModel.IndelModel.get_rates">get_rates</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>