<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>asymmetree.genome.GenomeSimulation API documentation</title>
<meta name="description" content="Simulation of genomes." />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>asymmetree.genome.GenomeSimulation</code></h1>
</header>
<section id="section-intro">
<p>Simulation of genomes.</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># -*- coding: utf-8 -*-

&#34;&#34;&#34;Simulation of genomes.&#34;&#34;&#34;

import os

from tralda.datastructures.Tree import Tree

import asymmetree.treeevolve as te
import asymmetree.seqevolve as se
from asymmetree.file_io.SeqFileIO import write_alignment, write_fasta
from asymmetree.tools.Sampling import Sampler


__author__ = &#39;David Schaller&#39;


class GenomeSimulator:
    
    def __init__(self, species_tree, outdir=None):
        &#34;&#34;&#34;
        Parameters
        ----------
        species_tree : Tree
            The species tree along which gene trees are simulated.
        outdir : str, optional
            The path to a directory into which the results of the simulation
            (serialized trees, one fasta file per genome, one &#39;true&#39; alignement
            per gene family). The default is None, in which case nothing is
            saved to file.
        &#34;&#34;&#34;
        
        if not isinstance(species_tree, Tree):
            raise TypeError(&#34;species tree must be of type &#39;Tree&#39;&#34;)
        
        self.S = species_tree
        self.outdir = outdir
        
        if self.outdir:
            self._check_outdir()
            self.S.serialize(self._path(&#39;species_tree.json&#39;), mode=&#39;json&#39;)
            
        self.true_gene_trees = []
        self.pruned_gene_trees = []
        
    
    def _check_outdir(self):
        
        if not os.path.exists(self.outdir):
            os.makedirs(self.outdir)
            
        elif os.path.exists(self.outdir) and not os.path.isdir(self.outdir):
            raise FileExistsError(&#34;&#39;{}&#39; is not a directory&#34;.format(self.outdir))
            
        for directory in (&#39;true_gene_trees&#39;, &#39;fasta_files&#39;, &#39;alignments&#39;):
            path = os.path.join(self.outdir, directory)
            if not os.path.exists(path):
                os.makedirs(path)
            
    
    def _path(self, *args):
        
        return os.path.join(self.outdir, *args)
            
    
    def simulate_gene_trees(self, n,
                            **kwargs):
        &#34;&#34;&#34;Simulate gene trees.
        
        Simulates the &#39;true gene trees&#39; which still contain the loss events as
        well as the &#39;pruned gene trees&#39; in which the loss branches are
        removed.
        
        Parameters
        ----------
        n : int
            Number of gene trees to be simulated.
        kwargs : optional
            See function &#39;gene_trees&#39; in &#39;treeevolve&#39; subpackage or 
            documentation.
        &#34;&#34;&#34;
        
        self.number_of_families = n
        
        self.true_gene_trees = te.gene_trees(self.S, n=n, **kwargs)
        
        self.pruned_gene_trees = [te.prune_losses(tree)
                                  for tree in self.true_gene_trees]
        
        # sequences should be emptied here if methods were called before
        if hasattr(self, &#39;sequence_dicts&#39;):
            self.sequence_dicts.clear()
        
        if self.outdir:
            for i in range(n):
                filename = self._path(&#39;true_gene_trees&#39;,
                                      &#39;gene_tree{}.json&#39;.format(i))
                self.true_gene_trees[i].serialize(filename, mode=&#39;json&#39;)
        
        return self.true_gene_trees, self.pruned_gene_trees
                
    
    def simulate_sequences(self, subst_model,
                           root_genome=None,
                           length_distr=(&#39;constant&#39;, 200),
                           min_length=10,
                           max_length=None,
                           write_fastas=True,
                           write_alignments=True,
                           **kwargs):
        &#34;&#34;&#34;Simulate sequences along the (pruned) gene trees.
        
        Parameters
        ----------
        subst_model : SubstModel
            Substitution model.
        root_genome : list, optional
            List of sequences for the roots of the gene trees, must contain the
            same number of str sequences as trees that were simulated. 
            The sequences must be compatible with the specified substitution 
            model (model_type=&#39;n&#39;/&#39;a&#39;). The default is None, in which case new
            sequences are generated at random.
        length_distr : tuple, optional
            Distribution of the length of the root sequences if root genome is
            not supplied. See documentation for available options.
        min_length : int, optional
            Minimal length at which the distribution of lengths is truncated;
            must be less than the mean of this distribution. The default is 10.
        max_length : int, optional
            Maximal length at which the distribution of lengths is truncated;
            must be less than the mean of this distribution. The default is
            None, in which case the distribution is not truncated.
        write_fastas : bool, optional
            If True and an output directory was speci ed, write the sequences
            (one file per species) into the directory &#39;fasta_files&#39; in the
            output directory. The default is True.
        write_alignments : bool, optional
            If True and an output directory was speci ed, write the true 
            alignments (one file per gene tree) into the directory &#39;alignments&#39;
            in the output directory. The default is True.
        kwargs : optional
            See class &#39;Evolver&#39; in the &#39;seqevolve&#39; subpackage or documentation.
        &#34;&#34;&#34;
        
        self.subst_model = subst_model
        
        if hasattr(self, &#39;sequence_dicts&#39;):
            self.sequence_dicts.clear()
        else:
            self.sequence_dicts = []
        
        if root_genome:
            if len(root_genome) != len(self.number_of_families):
                raise ValueError(&#39;no. of sequences in root genome does not&#39;\
                                 &#39;match no of gene families&#39;)
        else:
            if isinstance(length_distr, Sampler):
                self.sampler = length_distr
            else:
                self.sampler = Sampler(length_distr,
                                       minimum=min_length,
                                       maximum=max_length,
                                       discrete=True)
        
        evolver = se.Evolver(subst_model, **kwargs)
        
        for i in range(self.number_of_families):
            
            PGT = self.pruned_gene_trees[i]
            
            if root_genome:
                evolver.evolve_along_tree(PGT, start_seq=root_genome[i].upper())
                
            else:
                evolver.evolve_along_tree(PGT, start_length=self.sampler())
                
            self.sequence_dicts.append(evolver.sequences)
            
            # write one alignment file per gene family
            if self.outdir and write_alignments:
                self._write_alignment(i)
        
        # write one fasta file per species
        if self.outdir and write_fastas:
            self._write_fastas(include_inner=False)
            
            
    def _write_alignment(self, family_id):
        
        if not self.outdir:
            raise RuntimeError(&#39;no output directory specified for alignments&#39;)        
        
        alg_builder = se.AlignmentBuilder(self.pruned_gene_trees[family_id],
                                          self.sequence_dicts[family_id],
                                          self.subst_model.alphabet,
                                          include_inner=False)
        
        alignment = []
        for node, sequence in alg_builder.build().items():
            label = self._compose_label(node, family_id)
            alignment.append( (label, sequence) )
            
        basename = &#39;alignment{}.phylip&#39;.format(family_id)
        filename = self._path(&#39;alignments&#39;, basename)
        write_alignment(filename, alignment, alignment_format=&#39;phylip&#39;)
    
                
    def _write_fastas(self, include_inner=False):
        
        if not self.outdir:
            raise RuntimeError(&#39;no output directory specified for fasta files&#39;)
        
        # list of all species labels
        if include_inner:
            species = [v.label for v in self.S.preorder()]
        else:
            species = [v.label for v in self.S.preorder() if not v.children]
        
        # labeled sequences sorted by species
        sorted_seqs = {s: [] for s in species}
        
        for i in range(self.number_of_families):
            for node, evoseq in self.sequence_dicts[i].items():
                
                # skip inner nodes
                if not include_inner and node.children:
                    continue
                
                label = self._compose_label(node, i)
                sequence = self.subst_model.to_sequence(evoseq)
                sorted_seqs[node.reconc].append( (label, sequence) )
                
        for spec, sequences in sorted_seqs.items():
            basename = &#39;{}.f{}a&#39;.format(spec, self.subst_model.model_type)
            filename = self._path(&#39;fasta_files&#39;, basename)
            write_fasta(filename, sequences)
        
    
    def _compose_label(self, node, family_id):
        
        return &#39;fam{}gene{}spec{}&#39;.format(family_id, node.label, node.reconc)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="asymmetree.genome.GenomeSimulation.GenomeSimulator"><code class="flex name class">
<span>class <span class="ident">GenomeSimulator</span></span>
<span>(</span><span>species_tree, outdir=None)</span>
</code></dt>
<dd>
<div class="desc"><h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>species_tree</code></strong> :&ensp;<code>Tree</code></dt>
<dd>The species tree along which gene trees are simulated.</dd>
<dt><strong><code>outdir</code></strong> :&ensp;<code>str</code>, optional</dt>
<dd>The path to a directory into which the results of the simulation
(serialized trees, one fasta file per genome, one 'true' alignement
per gene family). The default is None, in which case nothing is
saved to file.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class GenomeSimulator:
    
    def __init__(self, species_tree, outdir=None):
        &#34;&#34;&#34;
        Parameters
        ----------
        species_tree : Tree
            The species tree along which gene trees are simulated.
        outdir : str, optional
            The path to a directory into which the results of the simulation
            (serialized trees, one fasta file per genome, one &#39;true&#39; alignement
            per gene family). The default is None, in which case nothing is
            saved to file.
        &#34;&#34;&#34;
        
        if not isinstance(species_tree, Tree):
            raise TypeError(&#34;species tree must be of type &#39;Tree&#39;&#34;)
        
        self.S = species_tree
        self.outdir = outdir
        
        if self.outdir:
            self._check_outdir()
            self.S.serialize(self._path(&#39;species_tree.json&#39;), mode=&#39;json&#39;)
            
        self.true_gene_trees = []
        self.pruned_gene_trees = []
        
    
    def _check_outdir(self):
        
        if not os.path.exists(self.outdir):
            os.makedirs(self.outdir)
            
        elif os.path.exists(self.outdir) and not os.path.isdir(self.outdir):
            raise FileExistsError(&#34;&#39;{}&#39; is not a directory&#34;.format(self.outdir))
            
        for directory in (&#39;true_gene_trees&#39;, &#39;fasta_files&#39;, &#39;alignments&#39;):
            path = os.path.join(self.outdir, directory)
            if not os.path.exists(path):
                os.makedirs(path)
            
    
    def _path(self, *args):
        
        return os.path.join(self.outdir, *args)
            
    
    def simulate_gene_trees(self, n,
                            **kwargs):
        &#34;&#34;&#34;Simulate gene trees.
        
        Simulates the &#39;true gene trees&#39; which still contain the loss events as
        well as the &#39;pruned gene trees&#39; in which the loss branches are
        removed.
        
        Parameters
        ----------
        n : int
            Number of gene trees to be simulated.
        kwargs : optional
            See function &#39;gene_trees&#39; in &#39;treeevolve&#39; subpackage or 
            documentation.
        &#34;&#34;&#34;
        
        self.number_of_families = n
        
        self.true_gene_trees = te.gene_trees(self.S, n=n, **kwargs)
        
        self.pruned_gene_trees = [te.prune_losses(tree)
                                  for tree in self.true_gene_trees]
        
        # sequences should be emptied here if methods were called before
        if hasattr(self, &#39;sequence_dicts&#39;):
            self.sequence_dicts.clear()
        
        if self.outdir:
            for i in range(n):
                filename = self._path(&#39;true_gene_trees&#39;,
                                      &#39;gene_tree{}.json&#39;.format(i))
                self.true_gene_trees[i].serialize(filename, mode=&#39;json&#39;)
        
        return self.true_gene_trees, self.pruned_gene_trees
                
    
    def simulate_sequences(self, subst_model,
                           root_genome=None,
                           length_distr=(&#39;constant&#39;, 200),
                           min_length=10,
                           max_length=None,
                           write_fastas=True,
                           write_alignments=True,
                           **kwargs):
        &#34;&#34;&#34;Simulate sequences along the (pruned) gene trees.
        
        Parameters
        ----------
        subst_model : SubstModel
            Substitution model.
        root_genome : list, optional
            List of sequences for the roots of the gene trees, must contain the
            same number of str sequences as trees that were simulated. 
            The sequences must be compatible with the specified substitution 
            model (model_type=&#39;n&#39;/&#39;a&#39;). The default is None, in which case new
            sequences are generated at random.
        length_distr : tuple, optional
            Distribution of the length of the root sequences if root genome is
            not supplied. See documentation for available options.
        min_length : int, optional
            Minimal length at which the distribution of lengths is truncated;
            must be less than the mean of this distribution. The default is 10.
        max_length : int, optional
            Maximal length at which the distribution of lengths is truncated;
            must be less than the mean of this distribution. The default is
            None, in which case the distribution is not truncated.
        write_fastas : bool, optional
            If True and an output directory was speci ed, write the sequences
            (one file per species) into the directory &#39;fasta_files&#39; in the
            output directory. The default is True.
        write_alignments : bool, optional
            If True and an output directory was speci ed, write the true 
            alignments (one file per gene tree) into the directory &#39;alignments&#39;
            in the output directory. The default is True.
        kwargs : optional
            See class &#39;Evolver&#39; in the &#39;seqevolve&#39; subpackage or documentation.
        &#34;&#34;&#34;
        
        self.subst_model = subst_model
        
        if hasattr(self, &#39;sequence_dicts&#39;):
            self.sequence_dicts.clear()
        else:
            self.sequence_dicts = []
        
        if root_genome:
            if len(root_genome) != len(self.number_of_families):
                raise ValueError(&#39;no. of sequences in root genome does not&#39;\
                                 &#39;match no of gene families&#39;)
        else:
            if isinstance(length_distr, Sampler):
                self.sampler = length_distr
            else:
                self.sampler = Sampler(length_distr,
                                       minimum=min_length,
                                       maximum=max_length,
                                       discrete=True)
        
        evolver = se.Evolver(subst_model, **kwargs)
        
        for i in range(self.number_of_families):
            
            PGT = self.pruned_gene_trees[i]
            
            if root_genome:
                evolver.evolve_along_tree(PGT, start_seq=root_genome[i].upper())
                
            else:
                evolver.evolve_along_tree(PGT, start_length=self.sampler())
                
            self.sequence_dicts.append(evolver.sequences)
            
            # write one alignment file per gene family
            if self.outdir and write_alignments:
                self._write_alignment(i)
        
        # write one fasta file per species
        if self.outdir and write_fastas:
            self._write_fastas(include_inner=False)
            
            
    def _write_alignment(self, family_id):
        
        if not self.outdir:
            raise RuntimeError(&#39;no output directory specified for alignments&#39;)        
        
        alg_builder = se.AlignmentBuilder(self.pruned_gene_trees[family_id],
                                          self.sequence_dicts[family_id],
                                          self.subst_model.alphabet,
                                          include_inner=False)
        
        alignment = []
        for node, sequence in alg_builder.build().items():
            label = self._compose_label(node, family_id)
            alignment.append( (label, sequence) )
            
        basename = &#39;alignment{}.phylip&#39;.format(family_id)
        filename = self._path(&#39;alignments&#39;, basename)
        write_alignment(filename, alignment, alignment_format=&#39;phylip&#39;)
    
                
    def _write_fastas(self, include_inner=False):
        
        if not self.outdir:
            raise RuntimeError(&#39;no output directory specified for fasta files&#39;)
        
        # list of all species labels
        if include_inner:
            species = [v.label for v in self.S.preorder()]
        else:
            species = [v.label for v in self.S.preorder() if not v.children]
        
        # labeled sequences sorted by species
        sorted_seqs = {s: [] for s in species}
        
        for i in range(self.number_of_families):
            for node, evoseq in self.sequence_dicts[i].items():
                
                # skip inner nodes
                if not include_inner and node.children:
                    continue
                
                label = self._compose_label(node, i)
                sequence = self.subst_model.to_sequence(evoseq)
                sorted_seqs[node.reconc].append( (label, sequence) )
                
        for spec, sequences in sorted_seqs.items():
            basename = &#39;{}.f{}a&#39;.format(spec, self.subst_model.model_type)
            filename = self._path(&#39;fasta_files&#39;, basename)
            write_fasta(filename, sequences)
        
    
    def _compose_label(self, node, family_id):
        
        return &#39;fam{}gene{}spec{}&#39;.format(family_id, node.label, node.reconc)</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="asymmetree.genome.GenomeSimulation.GenomeSimulator.simulate_gene_trees"><code class="name flex">
<span>def <span class="ident">simulate_gene_trees</span></span>(<span>self, n, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Simulate gene trees.</p>
<p>Simulates the 'true gene trees' which still contain the loss events as
well as the 'pruned gene trees' in which the loss branches are
removed.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>n</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of gene trees to be simulated.</dd>
<dt><strong><code>kwargs</code></strong> :&ensp;<code>optional</code></dt>
<dd>See function 'gene_trees' in 'treeevolve' subpackage or
documentation.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def simulate_gene_trees(self, n,
                        **kwargs):
    &#34;&#34;&#34;Simulate gene trees.
    
    Simulates the &#39;true gene trees&#39; which still contain the loss events as
    well as the &#39;pruned gene trees&#39; in which the loss branches are
    removed.
    
    Parameters
    ----------
    n : int
        Number of gene trees to be simulated.
    kwargs : optional
        See function &#39;gene_trees&#39; in &#39;treeevolve&#39; subpackage or 
        documentation.
    &#34;&#34;&#34;
    
    self.number_of_families = n
    
    self.true_gene_trees = te.gene_trees(self.S, n=n, **kwargs)
    
    self.pruned_gene_trees = [te.prune_losses(tree)
                              for tree in self.true_gene_trees]
    
    # sequences should be emptied here if methods were called before
    if hasattr(self, &#39;sequence_dicts&#39;):
        self.sequence_dicts.clear()
    
    if self.outdir:
        for i in range(n):
            filename = self._path(&#39;true_gene_trees&#39;,
                                  &#39;gene_tree{}.json&#39;.format(i))
            self.true_gene_trees[i].serialize(filename, mode=&#39;json&#39;)
    
    return self.true_gene_trees, self.pruned_gene_trees</code></pre>
</details>
</dd>
<dt id="asymmetree.genome.GenomeSimulation.GenomeSimulator.simulate_sequences"><code class="name flex">
<span>def <span class="ident">simulate_sequences</span></span>(<span>self, subst_model, root_genome=None, length_distr=('constant', 200), min_length=10, max_length=None, write_fastas=True, write_alignments=True, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>Simulate sequences along the (pruned) gene trees.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>subst_model</code></strong> :&ensp;<code>SubstModel</code></dt>
<dd>Substitution model.</dd>
<dt><strong><code>root_genome</code></strong> :&ensp;<code>list</code>, optional</dt>
<dd>List of sequences for the roots of the gene trees, must contain the
same number of str sequences as trees that were simulated.
The sequences must be compatible with the specified substitution
model (model_type='n'/'a'). The default is None, in which case new
sequences are generated at random.</dd>
<dt><strong><code>length_distr</code></strong> :&ensp;<code>tuple</code>, optional</dt>
<dd>Distribution of the length of the root sequences if root genome is
not supplied. See documentation for available options.</dd>
<dt><strong><code>min_length</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>Minimal length at which the distribution of lengths is truncated;
must be less than the mean of this distribution. The default is 10.</dd>
<dt><strong><code>max_length</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>Maximal length at which the distribution of lengths is truncated;
must be less than the mean of this distribution. The default is
None, in which case the distribution is not truncated.</dd>
<dt><strong><code>write_fastas</code></strong> :&ensp;<code>bool</code>, optional</dt>
<dd>If True and an output directory was speci ed, write the sequences
(one file per species) into the directory 'fasta_files' in the
output directory. The default is True.</dd>
<dt><strong><code>write_alignments</code></strong> :&ensp;<code>bool</code>, optional</dt>
<dd>If True and an output directory was speci ed, write the true
alignments (one file per gene tree) into the directory 'alignments'
in the output directory. The default is True.</dd>
<dt><strong><code>kwargs</code></strong> :&ensp;<code>optional</code></dt>
<dd>See class 'Evolver' in the 'seqevolve' subpackage or documentation.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def simulate_sequences(self, subst_model,
                       root_genome=None,
                       length_distr=(&#39;constant&#39;, 200),
                       min_length=10,
                       max_length=None,
                       write_fastas=True,
                       write_alignments=True,
                       **kwargs):
    &#34;&#34;&#34;Simulate sequences along the (pruned) gene trees.
    
    Parameters
    ----------
    subst_model : SubstModel
        Substitution model.
    root_genome : list, optional
        List of sequences for the roots of the gene trees, must contain the
        same number of str sequences as trees that were simulated. 
        The sequences must be compatible with the specified substitution 
        model (model_type=&#39;n&#39;/&#39;a&#39;). The default is None, in which case new
        sequences are generated at random.
    length_distr : tuple, optional
        Distribution of the length of the root sequences if root genome is
        not supplied. See documentation for available options.
    min_length : int, optional
        Minimal length at which the distribution of lengths is truncated;
        must be less than the mean of this distribution. The default is 10.
    max_length : int, optional
        Maximal length at which the distribution of lengths is truncated;
        must be less than the mean of this distribution. The default is
        None, in which case the distribution is not truncated.
    write_fastas : bool, optional
        If True and an output directory was speci ed, write the sequences
        (one file per species) into the directory &#39;fasta_files&#39; in the
        output directory. The default is True.
    write_alignments : bool, optional
        If True and an output directory was speci ed, write the true 
        alignments (one file per gene tree) into the directory &#39;alignments&#39;
        in the output directory. The default is True.
    kwargs : optional
        See class &#39;Evolver&#39; in the &#39;seqevolve&#39; subpackage or documentation.
    &#34;&#34;&#34;
    
    self.subst_model = subst_model
    
    if hasattr(self, &#39;sequence_dicts&#39;):
        self.sequence_dicts.clear()
    else:
        self.sequence_dicts = []
    
    if root_genome:
        if len(root_genome) != len(self.number_of_families):
            raise ValueError(&#39;no. of sequences in root genome does not&#39;\
                             &#39;match no of gene families&#39;)
    else:
        if isinstance(length_distr, Sampler):
            self.sampler = length_distr
        else:
            self.sampler = Sampler(length_distr,
                                   minimum=min_length,
                                   maximum=max_length,
                                   discrete=True)
    
    evolver = se.Evolver(subst_model, **kwargs)
    
    for i in range(self.number_of_families):
        
        PGT = self.pruned_gene_trees[i]
        
        if root_genome:
            evolver.evolve_along_tree(PGT, start_seq=root_genome[i].upper())
            
        else:
            evolver.evolve_along_tree(PGT, start_length=self.sampler())
            
        self.sequence_dicts.append(evolver.sequences)
        
        # write one alignment file per gene family
        if self.outdir and write_alignments:
            self._write_alignment(i)
    
    # write one fasta file per species
    if self.outdir and write_fastas:
        self._write_fastas(include_inner=False)</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="asymmetree.genome" href="index.html">asymmetree.genome</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="asymmetree.genome.GenomeSimulation.GenomeSimulator" href="#asymmetree.genome.GenomeSimulation.GenomeSimulator">GenomeSimulator</a></code></h4>
<ul class="">
<li><code><a title="asymmetree.genome.GenomeSimulation.GenomeSimulator.simulate_gene_trees" href="#asymmetree.genome.GenomeSimulation.GenomeSimulator.simulate_gene_trees">simulate_gene_trees</a></code></li>
<li><code><a title="asymmetree.genome.GenomeSimulation.GenomeSimulator.simulate_sequences" href="#asymmetree.genome.GenomeSimulation.GenomeSimulator.simulate_sequences">simulate_sequences</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>