<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>asymmetree.file_io.ProteinOrthoParser API documentation</title>
<meta name="description" content="Parser for ProteinOrtho files …" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>asymmetree.file_io.ProteinOrthoParser</code></h1>
</header>
<section id="section-intro">
<p>Parser for ProteinOrtho files.</p>
<p>Reads two different versions of ProteinOrtho result files (with/without
'-synteny' option) and builds a graph from the data (color &ndash;&gt;
species/filename).</p>
<h2 id="references">References</h2>
<ul>
<li>Marcus Lechner, Sven Findeiß, Lydia Steiner, Manja Marz, Peter F.
Stadler, Sonja J. Prohaska. Proteinortho: Detection of (Co-)orthologs
in large-scale analysis. BMC Bioinformatics 2011, 12:124</li>
</ul>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python"># -*- coding: utf-8 -*-

&#34;&#34;&#34; 
Parser for ProteinOrtho files.

Reads two different versions of ProteinOrtho result files (with/without
&#39;-synteny&#39; option) and builds a graph from the data (color --&gt; 
species/filename).

References:
    - Marcus Lechner, Sven Findeiß, Lydia Steiner, Manja Marz, Peter F.
      Stadler, Sonja J. Prohaska. Proteinortho: Detection of (Co-)orthologs
      in large-scale analysis. BMC Bioinformatics 2011, 12:124
&#34;&#34;&#34;

import networkx as nx   


__author__ = &#39;David Schaller&#39;


def parse_po_graph(filename):
    
    G = nx.Graph()
    
    with open(filename, &#39;r&#39;) as f:
        
        # skip the first two lines
        f.readline(); f.readline
        
        col_a, col_b = &#39;unknown&#39;, &#39;unknown&#39;
        line = f.readline().strip()
        
        while line:
            
            line = line.strip()
            
            # new color pair begins
            if line and line.startswith(&#39;#&#39;) and not line.startswith(&#39;# Scores:&#39;):
                colors = line.split()
                if len(colors) == 3:
                    col_a, col_b = colors[1], colors[2]
                
            elif line and not line.startswith(&#39;#&#39;):
                edge = line.split()
                id1 = &#39;{}_{}&#39;.format(col_a, edge[0])
                id2 = &#39;{}_{}&#39;.format(col_b, edge[1])
                if id1 not in G:
                    G.add_node(id1, color=col_a)
                if id2 not in G:
                    G.add_node(id2, color=col_b)
                    
                # file type 1 with simscore (synteny was activated)
                if len(edge) == 8:
                    G.add_edge(id1, id2,
                               evalue_ab = float(edge[2]), 
                               bitscore_ab = float(edge[3]),
                               evalue_ba = float(edge[4]),
                               bitscore_ba = float(edge[5]),
                               same_strand = int(edge[6]),
                               simscore = float(edge[7]))
                
                # file type 2 without simscore
                elif len(edge) == 6:
                    G.add_edge(id1, id2,
                               evalue_ab = float(edge[2]), 
                               bitscore_ab = float(edge[3]),
                               evalue_ba = float(edge[4]),
                               bitscore_ba = float(edge[5]))
                else:
                    print(&#39;Check file format!&#39;, edge)
                    continue
                
            line = f.readline()
            
    return G


def parse_best_match_candidates(filename):
    
    candidates = {}
    
#    # extract the species from filename
#    species_match = re.search(r&#34;([^/]+)\.bm_candidates$&#34;, filename)
#    if species_match:
#        species = species_match.group(1)
#    else:
#        raise ValueError(&#34;could not extract species from filename &#39;{}&#39;&#34;.format(filename))
    
    with open(filename, &#39;r&#39;) as f:
        
        line = f.readline()
        
        while line:
            
            line = line.strip()
            if not line:
                line = f.readline()
                continue
            
            data = line.split()
            
            if len(data) != 6:
                print(&#39;Check file format: {}, line {}&#39;.format(filename, line))
                continue
                
            color      = data[0]
            query_id   = data[1]
            subject_id = data[2]
            e_value    = float(data[3])
            bitscore   = float(data[4])
            identity   = float(data[5]) / 100
            
            if subject_id not in candidates:
                candidates[subject_id] = {}
            if color not in candidates[subject_id]:
                candidates[subject_id][color] = []
            candidates[subject_id][color].append( (color, query_id, e_value, bitscore, identity) )
                
            line = f.readline()
            
    return candidates


def parse_best_matches(filename):
    
    G = nx.DiGraph()
    
    with open(filename, &#39;r&#39;) as f:
        
        line = f.readline()
        
        while line:
            
            line = line.strip().split()
            
            if len(line) == 4:
                
                col1, col2 = line[2], line[3]
                id1 = &#39;{}_{}&#39;.format(col1, line[0])
                id2 = &#39;{}_{}&#39;.format(col2, line[1])
                if id1 not in G:
                    G.add_node(id1, color=col1)
                if id2 not in G:
                    G.add_node(id2, color=col2)
                G.add_edge(id1, id2)
            
            line = f.readline()
            
    return G</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="asymmetree.file_io.ProteinOrthoParser.parse_best_match_candidates"><code class="name flex">
<span>def <span class="ident">parse_best_match_candidates</span></span>(<span>filename)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def parse_best_match_candidates(filename):
    
    candidates = {}
    
#    # extract the species from filename
#    species_match = re.search(r&#34;([^/]+)\.bm_candidates$&#34;, filename)
#    if species_match:
#        species = species_match.group(1)
#    else:
#        raise ValueError(&#34;could not extract species from filename &#39;{}&#39;&#34;.format(filename))
    
    with open(filename, &#39;r&#39;) as f:
        
        line = f.readline()
        
        while line:
            
            line = line.strip()
            if not line:
                line = f.readline()
                continue
            
            data = line.split()
            
            if len(data) != 6:
                print(&#39;Check file format: {}, line {}&#39;.format(filename, line))
                continue
                
            color      = data[0]
            query_id   = data[1]
            subject_id = data[2]
            e_value    = float(data[3])
            bitscore   = float(data[4])
            identity   = float(data[5]) / 100
            
            if subject_id not in candidates:
                candidates[subject_id] = {}
            if color not in candidates[subject_id]:
                candidates[subject_id][color] = []
            candidates[subject_id][color].append( (color, query_id, e_value, bitscore, identity) )
                
            line = f.readline()
            
    return candidates</code></pre>
</details>
</dd>
<dt id="asymmetree.file_io.ProteinOrthoParser.parse_best_matches"><code class="name flex">
<span>def <span class="ident">parse_best_matches</span></span>(<span>filename)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def parse_best_matches(filename):
    
    G = nx.DiGraph()
    
    with open(filename, &#39;r&#39;) as f:
        
        line = f.readline()
        
        while line:
            
            line = line.strip().split()
            
            if len(line) == 4:
                
                col1, col2 = line[2], line[3]
                id1 = &#39;{}_{}&#39;.format(col1, line[0])
                id2 = &#39;{}_{}&#39;.format(col2, line[1])
                if id1 not in G:
                    G.add_node(id1, color=col1)
                if id2 not in G:
                    G.add_node(id2, color=col2)
                G.add_edge(id1, id2)
            
            line = f.readline()
            
    return G</code></pre>
</details>
</dd>
<dt id="asymmetree.file_io.ProteinOrthoParser.parse_po_graph"><code class="name flex">
<span>def <span class="ident">parse_po_graph</span></span>(<span>filename)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def parse_po_graph(filename):
    
    G = nx.Graph()
    
    with open(filename, &#39;r&#39;) as f:
        
        # skip the first two lines
        f.readline(); f.readline
        
        col_a, col_b = &#39;unknown&#39;, &#39;unknown&#39;
        line = f.readline().strip()
        
        while line:
            
            line = line.strip()
            
            # new color pair begins
            if line and line.startswith(&#39;#&#39;) and not line.startswith(&#39;# Scores:&#39;):
                colors = line.split()
                if len(colors) == 3:
                    col_a, col_b = colors[1], colors[2]
                
            elif line and not line.startswith(&#39;#&#39;):
                edge = line.split()
                id1 = &#39;{}_{}&#39;.format(col_a, edge[0])
                id2 = &#39;{}_{}&#39;.format(col_b, edge[1])
                if id1 not in G:
                    G.add_node(id1, color=col_a)
                if id2 not in G:
                    G.add_node(id2, color=col_b)
                    
                # file type 1 with simscore (synteny was activated)
                if len(edge) == 8:
                    G.add_edge(id1, id2,
                               evalue_ab = float(edge[2]), 
                               bitscore_ab = float(edge[3]),
                               evalue_ba = float(edge[4]),
                               bitscore_ba = float(edge[5]),
                               same_strand = int(edge[6]),
                               simscore = float(edge[7]))
                
                # file type 2 without simscore
                elif len(edge) == 6:
                    G.add_edge(id1, id2,
                               evalue_ab = float(edge[2]), 
                               bitscore_ab = float(edge[3]),
                               evalue_ba = float(edge[4]),
                               bitscore_ba = float(edge[5]))
                else:
                    print(&#39;Check file format!&#39;, edge)
                    continue
                
            line = f.readline()
            
    return G</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="asymmetree.file_io" href="index.html">asymmetree.file_io</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="asymmetree.file_io.ProteinOrthoParser.parse_best_match_candidates" href="#asymmetree.file_io.ProteinOrthoParser.parse_best_match_candidates">parse_best_match_candidates</a></code></li>
<li><code><a title="asymmetree.file_io.ProteinOrthoParser.parse_best_matches" href="#asymmetree.file_io.ProteinOrthoParser.parse_best_matches">parse_best_matches</a></code></li>
<li><code><a title="asymmetree.file_io.ProteinOrthoParser.parse_po_graph" href="#asymmetree.file_io.ProteinOrthoParser.parse_po_graph">parse_po_graph</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>